migrate_abilities.rb

```diff
@@ -0,0 +1,290 @@
+# frozen_string_literal: true
+# rubocop:disable all
+
+require_relative 'models'
+
+Dir['./models/**/*.rb'].sort.each { |file| require file }
+
+Sequel.extension :pg_json_ops
+require_relative 'lib/engine'
+
+# https://stackoverflow.com/a/13634446
+class Array
+  def in_groups(number)
+    group_size = size / number
+    leftovers = size % number
+
+    groups = []
+    start = 0
+    number.times do |index|
+      length = group_size + (leftovers > 0 && leftovers > index ? 1 : 0)
+      groups << slice(start, length)
+      start += length
+    end
+
+    groups
+  end
+end
+
+def switch_actions(actions, first, second)
+  first_idx = actions.index(first)
+  second_idx = actions.index(second)
+
+  id = second['id']
+  second['id'] = first['id']
+  first['id'] = id
+
+  actions[first_idx] = second
+  actions[second_idx] = first
+  return [first, second]
+end
+
+# Returns either the actions that are modified inplace, or nil if inserted/deleted
+def repair(game, original_actions, actions, broken_action)
+  optionalish_actions = %w[message buy_company]
+  action_idx = actions.index(broken_action)
+  action = broken_action['original_id'] || broken_action['id']
+  puts "http://18xx.games/game/#{game.id}?action=#{action}"
+  puts game.active_step
+  prev_actions = actions[0..action_idx - 1]
+  prev_action = prev_actions[prev_actions.rindex { |a| !optionalish_actions.include?(a['type']) }]
+  next_actions = actions[action_idx + 1..]
+  next_action = next_actions.find { |a| !optionalish_actions.include?(a['type']) }
+  puts broken_action
+
+  if broken_action['type'] == 'pass' &&
+     (((next_action['type'] == 'run_routes') && game.active_step.is_a?(Engine::Step::Route)) ||
+      ((next_action['type'] == 'dividend') && game.active_step.is_a?(Engine::Step::Dividend)) ||
+      ((next_action['type'] == 'buy_train') && game.active_step.is_a?(Engine::Step::BuyTrain)))
+    actions.delete(broken_action)
+    return
+
+  elsif game.active_step.is_a?(Engine::Step::Route) &&
+     broken_action['type'] == 'place_token' &&
+     prev_action['type'] == 'pass'
+    actions.delete(prev_action)
+    return
+
+  elsif (company = game.company_by_id(broken_action['entity'])) &&
```

Comment:
this `elsif` branch does a lot of the work for fixing up 18Chesapeake games where people bought/used C-P and B&S at the wrong time of the turn by shifting actions around where possible (i.e., if in the actual game they bought the company in after running routes and used it immediately, but they had enough money at the track step, this will move those actions back to the track step)

Reply:
Toby told me not to bother with this approach on Slack but I was already in too deep and wanted to solve it and make it work, even if we end up not using it 😝 

On one hand, 1662 games is a lot of games to pin, but on the other hand 1356 is definitely a lot of games to migrate on the prod server. 🤷 

Just using the other branches in this `repair()` does handle some cases where it's just a matter of deleting or inserting a pass action

Reply:
Also eventually I want to get some of these changes (like the forking support) into the  main migrate script but I've been kicking that can down the road

Reply:
> On one hand, 1662 games is a lot of games to pin, but on the other hand 1356 is definitely a lot of games to migrate on the prod server. 🤷

Solution: we'll be able to proxy a connection to the DB from my machine to run the migration logic locally